File: Wechat_chatroom_helper_android/app/src/main/kotlin/com/zdy/project/wechat_chatroom_helper/ui/helper/avatar/AvatarMaker.java
package com.zdy.project.wechat_chatroom_helper.ui.helper.avatar;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;

import utils.AppSaveInfoUtils;

/**
 * Created by Mr.Zdy on 2018/3/2.
 */

public class AvatarMaker {


    //自造群消息助手头像
    public static void makeChatRoomBitmap(Canvas canvas, Paint paint, int size, Bitmap drawable) {
        Bitmap whiteMask = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);

        whiteMask.eraseColor(Color.WHITE);

        drawable = Bitmap.createScaledBitmap(drawable, size / 2, size / 2, false).copy(Bitmap.Config.ARGB_8888, false);

        //生成图
        Bitmap raw = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);

        Canvas bitmapCanvas = new Canvas(raw);

        //绘制logo
        bitmapCanvas.drawBitmap(drawable, size / 4, size / 4, paint);

        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));

        //给logo染色
        bitmapCanvas.drawBitmap(whiteMask, 0, 0, paint);

        paint.setXfermode(null);

        if (AppSaveInfoUtils.INSTANCE.isCircleAvatarInfo()) {
            paint.setColor(0xFF12B7F6);
            canvas.drawCircle(size / 2, size / 2, size / 2, paint);
        } else {
            canvas.drawColor(0xFF12B7F6);
        }

        canvas.drawBitmap(raw, 0, 0, paint);

    }

    //自造公众号助手头像
    public static void makeOfficialBitmap(Canvas canvas, Paint paint, int size) {
        Bitmap whiteMask = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);

        whiteMask.eraseColor(Color.WHITE);

        Bitmap drawable = Bitmap.createBitmap(size / 2, size / 2, Bitmap.Config.ARGB_8888);

        Canvas logoCanvas = new Canvas(drawable);

        paint.setStrokeWidth(size / 20);

        paint.setStyle(Paint.Style.STROKE);

        paint.setColor(0xFF9F289F);

        logoCanvas.drawCircle(size / 8 + size / 20, size / 4, size / 8, paint);
        logoCanvas.drawCircle(size / 2 - size / 8 - size / 20, size / 4, size / 8, paint);

        //生成图
        Bitmap raw = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888);

        Canvas bitmapCanvas = new Canvas(raw);

        //绘制logo
        bitmapCanvas.drawBitmap(drawable, size / 4, size / 4, paint);

        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));

        //给logo染色
        bitmapCanvas.drawBitmap(whiteMask, 0, 0, paint);

        paint.setXfermode(null);

        if (AppSaveInfoUtils.INSTANCE.isCircleAvatarInfo()) {
            paint.setColor(0xFFF5CB00);
            paint.setStrokeWidth(0);
            paint.setStyle(Paint.Style.FILL_AND_STROKE);
            canvas.drawCircle(size / 2, size / 2, size / 2, paint);
        } else {
            canvas.drawColor(0xFFF5CB00);
        }

        canvas.drawBitmap(raw, 0, 0, paint);

    }

}

File: Wechat_chatroom_helper_android/app/src/main/kotlin/com/zdy/project/wechat_chatroom_helper/ui/helper/avatar/AvatarMaker.kt
package com.zdy.project.wechat_chatroom_helper.ui.helper.avatar

import android.content.Context
import android.graphics.*
import com.zdy.project.wechat_chatroom_helper.Constants.Drawable_String_Chatroom_Avatar

import utils.AppSaveInfoUtils

/**
 * Created by Mr.Zdy on 2018/3/2.
 */

object AvatarMaker {

    private val AVATAR_BLUE = 0xFF12B7F6.toInt()
    private val AVATAR_AMBER = 0xFFF5CB00.toInt()

    //自造群消息助手头像
    fun makeChatRoomBitmap(context: Context, canvas: Canvas, paint: Paint) {

        val iconSize = canvas.width
        val contentSize = canvas.width / 2

        //創建内部内容區域，尺寸為icon尺寸的一半
        val drawable = with(context) {
            val identifier = context.resources.getIdentifier(Drawable_String_Chatroom_Avatar, "drawable", context.packageName)
            val rawDrawable = BitmapFactory.decodeResource(context.resources, identifier)
            return@with Bitmap.createScaledBitmap(rawDrawable, contentSize, contentSize, false)
                    .copy(Bitmap.Config.ARGB_8888, false)
        }


        val whiteMask = Bitmap.createBitmap(iconSize, iconSize, Bitmap.Config.ARGB_8888)
        whiteMask.eraseColor(Color.WHITE)

        //生成图
        val raw = Bitmap.createBitmap(iconSize, iconSize, Bitmap.Config.ARGB_8888)

        val bitmapCanvas = Canvas(raw)

        //绘制logo
        bitmapCanvas.drawBitmap(drawable, (contentSize / 2).toFloat(), (contentSize / 2).toFloat(), paint)

        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)

        //给logo染色
        bitmapCanvas.drawBitmap(whiteMask, 0f, 0f, paint)

        paint.xfermode = null

        if (AppSaveInfoUtils.isCircleAvatarInfo()) {
            paint.color = AVATAR_BLUE
            canvas.drawCircle(contentSize.toFloat(), contentSize.toFloat(), contentSize.toFloat(), paint)
        } else {
            canvas.drawColor(AVATAR_BLUE)
        }

        canvas.drawBitmap(raw, 0f, 0f, paint)

    }

    //自造公众号助手头像
    fun makeOfficialBitmap(canvas: Canvas, paint: Paint, size: Int) {
        val whiteMask = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        whiteMask.eraseColor(Color.WHITE)

        val logoDrawable = Bitmap.createBitmap(size / 2, size / 2, Bitmap.Config.ARGB_8888)
        val logoCanvas = Canvas(logoDrawable)
        paint.strokeWidth = (size / 20).toFloat()
        paint.style = Paint.Style.STROKE
        paint.color = -0x60d761//随机颜色
        logoCanvas.drawCircle((size / 8 + size / 20).toFloat(), (size / 4).toFloat(), (size / 8).toFloat(), paint)
        logoCanvas.drawCircle((size / 2 - size / 8 - size / 20).toFloat(), (size / 4).toFloat(), (size / 8).toFloat(), paint)

        //生成图
        val raw = Bitmap.createBitmap(size, size, Bitmap.Config.ARGB_8888)
        val bitmapCanvas = Canvas(raw)

        //绘制logo
        bitmapCanvas.drawBitmap(logoDrawable, (size / 4).toFloat(), (size / 4).toFloat(), paint)

        paint.xfermode = PorterDuffXfermode(PorterDuff.Mode.SRC_IN)

        //给logo染色
        bitmapCanvas.drawBitmap(whiteMask, 0f, 0f, paint)

        paint.xfermode = null

        if (AppSaveInfoUtils.isCircleAvatarInfo()) {
            paint.color = AVATAR_AMBER
            paint.strokeWidth = 0f
            paint.style = Paint.Style.FILL_AND_STROKE
            canvas.drawCircle((size / 2).toFloat(), (size / 2).toFloat(), (size / 2).toFloat(), paint)
        } else {
            canvas.drawColor(AVATAR_AMBER)
        }

        canvas.drawBitmap(raw, 0f, 0f, paint)

    }

}
